src/features/broker-protection/actions/fill-form.js

```diff
@@ -0,0 +1,60 @@
+import { getElement } from '../utils.js'
+import { ErrorResponse, SuccessResponse } from '../types.js'
+
+/**
+ * Takes the solved captcha token and injects it into the page to solve the captcha
+ *
+ * @param action
+ * @return {import('../types.js').ActionResponse}
+ */
+export function fillForm (action, userData) {
+    const form = getElement(document, action.selector)
+    if (!form) return new ErrorResponse({ actionID: action.id, message: 'missing form' })
+
+    // fill out form for each step
+    for (const element of action.elements) {
+        // get the correct field of the form
+        const inputElem = getElement(form, element.selector)
+        // this works for IDs (i.e. #url wouldb e form.elements['url'])
+        // let inputElem = form.elements[element.selector]
+        // find the correct userData to put in the form
+        if (inputElem) {
+            // @ts-expect-error - double check if this is strict enough
+            // todo: determine if this requires any events to be dispatched also
+            setValueForInput(inputElem, userData[element.type])
+        }
+    }
+
+    return new SuccessResponse({ actionID: action.id, actionType: action.actionType, response: null })
+}
+
+/**
+ * NOTE: This code comes from Autofill, the reasoning is to make React autofilling work on Chrome and Safari.
+ *
+ * Ensures the value is set properly and dispatches events to simulate real user action
+ *
+ * @param {HTMLInputElement} el
+ * @param {string} val
+ * @return {boolean}
+ */
+function setValueForInput (el, val) {
+    el.dispatchEvent(new Event('keydown', { bubbles: true }))
+
+    // Access the original setter (needed to bypass React's implementation on mobile)
+    // @ts-expect-error - Object will not be undefined on this case
+    const originalSet = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value').set
```

Comment:
we could also capture this setter in captured-globals, so it's even safer

Reply:
@muodov Yep I agree - due the isolated nature of the rest of this code, I'd prefer to come back and move this after the merge, so you think that's ok?